policy_agent/policy_agent.py

```diff
@@ -0,0 +1,354 @@
+
+    def process_resource(self, action, kind, target):
+        """
+        Takes a target object and an action and updates internal resource pools
+        :param action: String of ["ADDED", "DELETED", "MODIFIED"] (returned by api watch)
+        :param kind: String of ["Pod", "Namespace", "Service", "Endpoints"]
+        :param target: json dict of resource info
+        """
+        # Determine Resource Pool 
+        if kind == "Pod":
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == "Service":
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == "Endpoints":
+            resource_pool = self.endpoints
+            obj = Endopints(target)
+
+        elif kind == "Namespace":
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        else:
+            _log.error("resource %s not Pod, Service, or Namespace" % kind)
+            sys.exit(1)
+
+        target_key = obj.key
+
+        if action == "ADDED":
+            if target_key not in resource_pool:
+                resource_pool[target_key] = obj
+                _log.info("%s added to Calico store" % target_key)
+
+            else:
+                _log.error("Tried to Add %s, but %s already in bin" %
+                           (obj, target_key))
+
+        elif action == "DELETED":
+            if target_key in resource_pool:
+                del resource_pool[target_key]
+                _log.info("%s deleted from Calico store" % target_key)
+
+            else:
+                _log.error(
+                    "Tried to Delete %s, but %s was not in bin" %
+                    (obj, target_key))
+
+        elif action == "MODIFIED":
+            if target_key in resource_pool:
+                _log.info("Updating %s\n%s\n=====>\n%s" %
+                          (target_key, resource_pool[target_key], obj))
+                resource_pool[target_key] = obj
+
+            else:
+                _log.warning("Tried to Modify %s, but %s was not in bin. "
+                             "Treating as Addition" %
+                             (target_key, target_key))
+                self.process_resource(action="ADDED", kind=kind, target=target)
+
+    def resync(self):
+        """
+        Tells all resource objects to resync their profile information
+        """
+        for resource_pool in [self.pods, self.services, self.endpoints, self.namespaces]:
+            for resource_key in resource_pool:
+                resource_pool[resource_key].resync()
+
+
+class Resource():
+    """
+    Resource objects pull pertinent info from json blobs and maintain universal functions  
+    """
+
+    def __init__(self, json):
+        """
+        On init, each Resource saves the raw json, pulls necessary info (unique), defines a unique key identifier, and sets self.needs_resync to True
+        """
+        self.json = json
+        self.from_json(json)
```

Comment:
This is not quite the pattern I was thinking of with "from_json" - it works fine (especially since we're only ever creating one of these from json) but in the more generic case, you'd want something like this:

```
class Thing()
  def __init__(self, param1, param2, param3):
    self.p1 = param1
    self.p2 = param2
    self.p3 = param3

  @classmethod
  def from_json(cls, json):
    # Returns a new instance
    return Thing(jsondata["p1"], jsondata["p2"], jsondata["p3"]
```

Then you can do:
```
thing = Thing.from_json(json_data)

OR

thing = Thing(p1, p2, p3)
```